code/modules/power/smes.dm

```diff
@@ -112,55 +112,41 @@
 
 	//building and linking a terminal
 	if(istype(item, /obj/item/stack/cable_coil))
-		var/dir = get_dir(user,src)
-		if(dir & (dir-1))//we don't want diagonal click
-			return
-
-		if(terminal) //is there already a terminal ?
-			to_chat(user, span_warning("This SMES already has a power terminal!"))
-			return
-
-		if(!panel_open) //is the panel open ?
-			to_chat(user, span_warning("You must open the maintenance panel first!"))
-			return
-
-		var/turf/turf = get_turf(user)
-		if (turf.underfloor_accessibility < UNDERFLOOR_INTERACTABLE) //can we get to the underfloor?
-			to_chat(user, span_warning("You must first remove the floor plating!"))
-			return
-
-
-		var/obj/item/stack/cable_coil/cable = item
-		if(cable.get_amount() < 10)
-			to_chat(user, span_warning("You need more wires!"))
+		if(!can_place_terminal(user, item, silent = FALSE))
 			return
 
 		var/terminal_cable_layer
 		if(LAZYACCESS(params2list(params), RIGHT_CLICK))
 			var/choice = tgui_input_list(user, "Select Power Input Cable Layer", "Select Cable Layer", GLOB.cable_name_to_layer)
-			if(isnull(choice))
+			if(isnull(choice) \
+				|| !user.is_holding(item) \
+				|| !user.Adjacent(src) \
+				|| user.incapacitated() \
+				|| !can_place_terminal(user, item, silent = TRUE) \
+			)
 				return
 			terminal_cable_layer = GLOB.cable_name_to_layer[choice]
 
-		to_chat(user, span_notice("You start building the power terminal..."))
-		playsound(src.loc, 'sound/items/deconstruct.ogg', 50, TRUE)
+		user.visible_message(span_notice("[user.name] starts adding cables to [src]."))
+		balloon_alert(user, "adding cables...")
+		playsound(src, 'sound/items/deconstruct.ogg', 50, TRUE)
 
 		if(do_after(user, 2 SECONDS, target = src))
```

Comment:
would you mind turning this do_after() check into an early return? would make a lot of the new code you fixed up later on in this code-branch look a lot nicer